### ISSUE TITLE
Cross-encoder reranking: quality and performance improvements

### ISSUE THREAD

**Issue by XiaoConstantine** [XiaoConstantine/sgrep#1]
## Summary

Implementing two-stage retrieval with cross-encoder reranking to improve search quality, while addressing performance and quality challenges.

## Current State

The `feature/reranking` branch adds cross-encoder reranking using BGE-reranker-v2-m3 via llama.cpp server. This provides better precision by using a cross-encoder to re-score the top candidates from vector search.

### 1. Quality: Score Normalization Bug (Fixed)
- **Problem**: BGE reranker returns raw logits (e.g., -2.8, -11.0) but code assumed 0-1 range
- **Impact**: Relevant files like `pkg/errors/errors.go` dropped from #1 to #10 for "error handling" query
- **Solution**: Sigmoid normalization + RRF score fusion combining vector rank with reranker scores

### 2. Quality: Meta-queries fail (Fixed)
- **Problem**: Queries like "what does this repo do" didn't find README.md
- **Root cause**: Semantic gap between meta-question and chunk content
- **Solution**: Document-level embeddings (mean of chunk embeddings per file) with automatic fallback for meta-queries

### 3. Performance: Reranking bottleneck (Improved)
- **Problem**: Reranking took 99% of search time (~5s for 50 docs)
- **Solution**: 
  - Reduced candidate count (50 → 30)
  - Limited actual rerank to top 15 docs
  - Result: 2-6x speedup (1.4s → 250-680ms)

### New Features
- `file_embeddings` table with DiskANN vector index
- `FileEmbeddingStorer` interface for document-level search
- `IsMetaRepoQuery()` detection for automatic fallback
- RRF (Reciprocal Rank Fusion) combining vector + reranker scores
- Query intent detection (conceptual/definition/procedural/example)
- Code signal boosting based on file path patterns

### Files Modified
- `pkg/store/libsql.go` - file_embeddings table and methods
- `pkg/store/types.go` - FileEmbeddingStorer interface
- `pkg/index/index.go` - compute file embeddings during indexing
- `pkg/search/search.go` - RRF fusion, document-level search, performance limits

## Testing

```bash
# Quality test - error handling
./sgrep --rerank "error handling"
# Should return pkg/errors/errors.go as #1

# Meta-query test
./sgrep "what does this repo do"
# Should return README.md as #1

# Performance test
./sgrep -d --rerank "error handling"
# Should complete in <500ms
```

## Future Improvements

- [ ] Async/streaming reranking for better UX
- [ ] Smaller/faster reranker model option
- [ ] Configurable rerank limits via CLI flags
- [ ] Cache reranker results
